src/coredump/coredump.c

```diff
@@ -1086,40 +1086,65 @@ static int process_socket(int fd) {
                         goto finish;
                 }
 
-                /* The final zero-length datagram carries the file descriptor and tells us
+                /* The final zero-length datagram carries the file descriptors and tells us
                  * that we're done. */
                 if (n == 0) {
                         struct cmsghdr *found;
 
+                        if (first) {
+                                found = cmsg_find(&mh, SOL_SOCKET, SCM_RIGHTS, CMSG_LEN(sizeof(int)));
+                                if (found) {
+                                        /* This is the first message that carries file descriptors. Maybe
+                                         * there will be one more that actually contains array of two
+                                         * descriptors. */
+                                        assert(input_fd < 0);
+
+                                        input_fd = *CMSG_TYPED_DATA(found, int);
+                                        first = false;
+
+                                        continue;
+                                }
+
+                                /* This is zero length message but it doesn't carry a single descriptor.
+                                 * This is a protocol violation so let's bail out. */
+                                cmsg_close_all(&mh);
+                                r = log_error_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                        "Received zero length message with no file descriptor.");
```

Comment:
"no" is misleading. it might also carry more than one, in which case we'd hit this error path. maybe say "no or more than one"